LMB-326: Form options to separate predicates v2

JonasVanHoof opened the pull request:
Support form:options on a field but also form options as separate predicates e.g. formOption:conceptScheme.  

Other way of implementing https://github.com/lblod/ember-submission-form-fields/pull/188

JonasVanHoof commented on tests/dummy/public/test-forms/basic-fields/form.ttl:
```diff
@@ -249,6 +249,17 @@ ext:conceptSchemeSelectorField a form:Field ;
+    sh:name "Select with form options" ;
+    sh:order 270 ;
+    sh:path ext:conceptSchemeSelectorValue ;
+    formOption:conceptScheme exampleConceptSchemes:foo-bar-baz ;
+    formOption:searchEnabled true ;
```
Created prefix from the ext prefix

JonasVanHoof replied:
Not sure what uri to take for the `formOptions` prefix

Ref: https://chat.semte.ch/group/form-builder?msg=Q7ETZb9rqAE569WHF

JonasVanHoof commented on addon/components/rdf-input-fields/concept-scheme-selector.js:
```diff
@@ -33,17 +37,40 @@ export default class RdfInputFieldsConceptSchemeSelectorComponent extends InputF
+        return;
+      }
+
+      this.searchEnabled = isEnabled == '1' ? true : false;
```
Could convert it depending on datatype

Windvis replied:
You could also use the `Literal.toJS` util I think: http://linkeddata.github.io/rdflib.js/doc/classes/Literal.html#toJS

JonasVanHoof replied:
Yes nice one works great

JonasVanHoof commented on addon/components/rdf-input-fields/concept-scheme-selector.js:
```diff
@@ -33,17 +37,40 @@ export default class RdfInputFieldsConceptSchemeSelectorComponent extends InputF
+    let conceptScheme = null;
 
     if (!hasValidFieldOptions(this.args.field, ['conceptScheme'])) {
-      return;
-    }
+      const scheme = this.args.formStore.any(
+        this.args.field.uri,
+        FORM_OPTION('conceptScheme'),
+        undefined,
+        this.args.graphs.formGraph
+      );
+      if (!scheme) {
+        return;
+      }
 
-    const conceptScheme = new namedNode(fieldOptions.conceptScheme);
+      conceptScheme = scheme;
+    } else {
+      conceptScheme = new namedNode(fieldOptions.conceptScheme);
+    }
 
     /**
      * NOTE: Most forms are now implemented to have a default "true" behavior
      */
-    if (fieldOptions.searchEnabled !== undefined) {
+    if (!hasValidFieldOptions(this.args.field, ['searchEnabled'])) {
+      const isEnabled = this.args.formStore.any(
+        this.args.field.uri,
+        FORM_OPTION('searchEnabled'),
+        undefined,
+        this.args.graphs.formGraph
+      );
+      if (!isEnabled) {
+        return;
+      }
+
+      this.searchEnabled = isEnabled == '1' ? true : false;
+    } else {
       this.searchEnabled = fieldOptions.searchEnabled;
     }
```
This is a comparison of checking it in the component itself or using it through the base input component.
```js
  getOptionPredicates() {
    return {
      conceptScheme: EXT('conceptScheme'),
    };
  }

  loadOptions() {
    const metaGraph = this.args.graphs.metaGraph;
    const conceptScheme = this.findFieldOption('conceptScheme', 'node');

    if (!conceptScheme) {
      return;
    }

Rahien replied:
there are also more components where we need to add the new way of fetching options (by predicate) right? this PR is just an example for now?

Windvis replied:
@Rahien Yes, it only an example so we can compare the implementations and choose which option to go for.

Rahien commented on addon/components/rdf-input-fields/concept-scheme-selector.js:
```diff
@@ -33,17 +37,40 @@ export default class RdfInputFieldsConceptSchemeSelectorComponent extends InputF
     if (!hasValidFieldOptions(this.args.field, ['conceptScheme'])) {
-      return;
-    }
+      const scheme = this.args.formStore.any(
```
if we are going this route, i would switch the order of the checks, first checking for the predicate version as that will become the default and the old one will be deprecated.

JonasVanHoof replied:
updated it so it first checks the predicate and than the form:options if nothing was found for the predicate. Good one 👌🏻

JonasVanHoof commented:
@Rahien @Windvis  I updated all the input field components that use form:options. There are also components like the alert and the case-number that are also using these form:options. Should we also update them?

For the components that are updated at this moment: I added a getter for the field options. First did it inline just getting it from the store but it felt that the method was becoming to big. That's why I used the extra method :)

Windvis commented:
(FYI, I won't really have time to review this before next week (I don't work on fridays), so hopefully someone else can chime in 🙏. I think everything that doesn't increase the public API is mergeable since we can always refactor when needed.)

bfidlers commented:
Looks good, but I'm wondering if it wouldn't be helpful to write a helper function to parse the options? It's always pretty much the same thing you do, sometimes it is a bit different if it is a boolean or ... but should be easy enought to add this I think and would prevent a lot of similar code in future uses, plus makes it easier to use for other people.

JonasVanHoof commented:
Added a new page to the wiki where the possibilities are explained for the field options https://github.com/lblod/ember-submission-form-fields/wiki/Field-options

Windvis commented on tests/dummy/app/routes/application.js:
```diff
@@ -2,9 +2,9 @@ import Route from '@ember/routing/route';
 
 export default class ApplicationRoute extends Route {
-  // beforeModel() {
-  //   return mockRequestsWorker.start({
```
What was the issue again? We should get that resolved so you don't have to keep commenting this locally 😄

JonasVanHoof replied:
Haha I indeed commented this part everytime I wanted to edit something. Just ran the dummy app without commenting it and it did nothing strange! So you probably fixed it? Or mAgIc 😄

Windvis reviewed (commented):
Looks good to me! 👍  I would like a definite answer about the namespace URI in the internal chat before merging though 😄.

JonasVanHoof commented on tests/dummy/public/test-forms/basic-fields/form.ttl:
```diff
@@ -106,31 +107,45 @@ ext:conceptSchemeMultiSelectCheckboxesField a form:Field ;
+        form:grouping form:Bag ;
+        sh:resultMessage "Dit veld is verplicht."@nl;
+        sh:path ext:conceptSchemeMultiSelectCheckboxesValue] ;
+    fieldOption:conceptScheme exampleConceptSchemes:foo-bar-baz ;
+    fieldOption:orderBy qb:order ;
```
Alright will  update the exising ones 👌🏻

JonasVanHoof replied:
Updated and added the form:options way of adding configuration to a field as predicated in the wiki page
![image](https://github.com/lblod/ember-submission-form-fields/assets/36266973/df380caa-360c-45d0-98d2-d89978d66f9b)

JonasVanHoof commented:
We ended up choosing for the field options uri `<http://lblod.data.gift/vocabularies/field-options/>`
